src/lib/viewers/media/DashViewer.js

```diff
@@ -505,6 +506,28 @@ class DashViewer extends VideoBaseViewer {
                 this.handleDownloadError(error, downloadURL);
                 return;
             }
+
+            if (
+                normalizedShakaError.code === shaka.util.Error.Code.BAD_HTTP_STATUS &&
+                normalizedShakaError.data[1] === 401 // unauthorized error may be cuased by token expired
+            ) {
+                if (this.retryTokenCount >= MAX_RETRY_TOKEN) {
+                    error.details.error_message = 'Reach refreshing token limit for unauthorized error.';
+                } else {
+                    this.options
+                        .refreshToken()
+                        .then(newToken => {
+                            this.options.token = newToken;
+                            this.player.retryStreaming();
```

Comment:
Should we reset `retryTokenCount` on successful re-load?

Reply:
Successfully generating a new token doesn't mean that the new token is valid. If a user gives a problematic token generator function, we can always get a new invalid token, and it's still 401 error. Then we'll refresh it infinitely.